File: Examples/csharp/csharp.Tests/PointsTests.cs
using Elektronik.Common.Data.Pb;
using NUnit.Framework;
using System;
using System.IO;
using System.Linq;

namespace csharp.Tests
{
    public class PointsTests : TestsBase
    {
        private PointPb[] m_map;
        private ConnectionPb[] m_connections;

        [SetUp]
        public override void Setup()
        {
            base.Setup();
            m_map = Enumerable.Range(0, 5).Select(id => new PointPb()
            {
                Id = id,
                Message = $"{id}",
            }).ToArray();

            m_map[0].Position = new Vector3Pb() { X = 0, Y = .5 };
            m_map[1].Position = new Vector3Pb() { X = .5, Y = -.5 };
            m_map[2].Position = new Vector3Pb() { X = -.5, Y = -.5 };
            m_map[3].Position = new Vector3Pb() { X = -.5, Y = 0 };
            m_map[4].Position = new Vector3Pb() { X = .5, Y = 0 };

            m_connections = new[]
            {
                new ConnectionPb() { Id1 = m_map[0].Id, Id2 = m_map[1].Id, },
                new ConnectionPb() { Id1 = m_map[0].Id, Id2 = m_map[2].Id, },
                new ConnectionPb() { Id1 = m_map[2].Id, Id2 = m_map[4].Id, },
                new ConnectionPb() { Id1 = m_map[1].Id, Id2 = m_map[3].Id, },
                new ConnectionPb() { Id1 = m_map[3].Id, Id2 = m_map[4].Id, },
            };
        }

        [Test, Order(1)]
        public void Create()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Add,
                Points = new PacketPb.Types.Points(),
            };
            packet.Points.Data.Add(m_map);

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(2)]
        public void Update()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Points = new PacketPb.Types.Points(),
            };
            m_map[0].Position.Z += .5;
            m_map[2].Position.Z += .5;
            m_map[4].Position.Z += .5;

            packet.Points.Data.Add(m_map);

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(3)]
        public void UpdateConnections()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Points = new PacketPb.Types.Points(),
                Connections = new PacketPb.Types.Connections()
                {
                    Action = PacketPb.Types.Connections.Types.Action.Add,
                },
            };
            packet.Connections.Data.Add(m_connections);
            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(4)]
        public void RemoveConnections()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Points = new PacketPb.Types.Points(),
                Connections = new PacketPb.Types.Connections()
                {
                    Action = PacketPb.Types.Connections.Types.Action.Remove,
                },
            };
            packet.Connections.Data.Add(new[] { m_connections[0], m_connections[1] });
            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(5)]
        public void Remove()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Remove,
                Points = new PacketPb.Types.Points(),
            };

            packet.Points.Data.Add(new[] { m_map[1], m_map[3] });

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(6)]
        public void Clear()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Clear,
                Points = new PacketPb.Types.Points(),
            };

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }
    }
}

File: Examples/cxx/Tests/source/PointsTests.cpp
#include "PointsTests.h"

using namespace Elektronik::Common::Data::Pb;

void PointsTests::SetUp()
{
	TestsBase::SetUp();
    for (size_t i = 0; i < 5; ++i) {
        PointPb pt; 
        pt.set_id(i);
        pt.set_message(std::to_string(i));
        m_map.push_back(pt);
    }

    Vector3Pb pos; pos.set_x(0); pos.set_y(.5f);
    m_map[0].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(.5f); pos.set_y(-.5f);
    m_map[1].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(-.5f); pos.set_y(-.5f);
    m_map[2].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(-.5f);
    m_map[3].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(.5f);
    m_map[4].set_allocated_position(new Vector3Pb(pos));

    ConnectionPb connection; 
    
    connection.set_id1(m_map[0].id()); connection.set_id2(m_map[1].id());
    m_connections.push_back(connection);
    
    connection.set_id1(m_map[0].id()); connection.set_id2(m_map[2].id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[2].id()); connection.set_id2(m_map[4].id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[1].id()); connection.set_id2(m_map[3].id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[3].id()); connection.set_id2(m_map[4].id());
    m_connections.push_back(connection);
}

TEST_F(PointsTests, Create) {
    auto points = new PacketPb_Points();
    for (const auto pt : m_map) {
        auto newPt = points->add_data();
        newPt->CopyFrom(pt);
    }
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_add);
    packet.set_allocated_points(points);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(PointsTests, Update) {
    auto points = new PacketPb_Points();
    for (const auto pt : { m_map[2], m_map[4] }) {
        auto position = new Vector3Pb(pt.position());
        position->set_z(position->z() + .5f);
        auto newPt = points->add_data();
        newPt->CopyFrom(pt);
        newPt->set_allocated_position(position);
    }
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_points(points);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(PointsTests, UpdateConnections) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_points(new PacketPb_Points());

    auto connections = new PacketPb_Connections();
    connections->set_action(PacketPb_Connections_Action::PacketPb_Connections_Action_add);
    for (const auto c : m_connections) {
        auto c_ = connections->add_data();
        c_->CopyFrom(c);
    }
    packet.set_allocated_connections(connections);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(PointsTests, RemoveConnections) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_points(new PacketPb_Points());

    auto connections = new PacketPb_Connections();
    connections->set_action(PacketPb_Connections_Action::PacketPb_Connections_Action_remove);
    for (const auto& c : { m_connections[0], m_connections[1], }) {
        auto connection = connections->add_data();
        connection->CopyFrom(c);
    }
    packet.set_allocated_connections(connections);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(PointsTests, Remove) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_remove);
    auto pts2remove = new PacketPb_Points();
    for (const auto& pt : { m_map[1], m_map[3] }) {
        auto pt_ = pts2remove->add_data();
        pt_->set_id(pt.id());
    }
    packet.set_allocated_points(pts2remove);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(PointsTests, Clear) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_clear);
    packet.set_allocated_points(new PacketPb_Points());

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}